### PR TITLE
Feature/updated text bone

### PR DESCRIPTION
**Pull request by tsteinruecken**
Updated the HTML-Serializer in bones/textBone to support the new markup language (HTML5) used by the new text-edits and the new wysiwyg editor in Vi (Branch: feature/updatedTextbone)

(Merged from https://bitbucket.org/viur/server/pull-requests/95/feature-updatedtextbone/diff)

### CODE REVIEW

File: bones/textBone.py [hunk 1]
```diff
@@ -7,108 +7,193 @@
...
-					"td" : [ "colspan", "rowspan", "width", "height" ]+_attrsMargins+_attrsSpacing
-				},
-	"validStyles": ["font-weight","font-style","text-decoration","color", "display"], #List of CSS-Directives we allow
-	"singleTags": ["br","img", "hr"] # List of tags, which dont have a corresponding end tag
+	"validTags": [  # List of HTML-Tags which are valid
+		'b', 'a', 'i', 'u', 'span', 'div', 'p', 'img', 'ol', 'ul', 'li', 'abbr', 'sub', 'sup',
+		'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'table', 'thead', 'tbody', 'tfoot', 'tr', 'td', 'th', 'br',
+		'hr', 'strong', 'blockquote', 'em'],
+	"validAttrs": {  # Mapping of valid parameters for each tag (if a tag is not listed here: no parameters allowed)
+		"a": ["href", "target", "title"],
```
Comment (skoegl) [inline bones/textBone.py]: I would like to see 'rel' in validAttrs for a.

target="blank" should be used in conjunction with rel="noopener noreferrer".

Here more into the subject:
1. https://developers.google.com/web/tools/lighthouse/audits/noopener
2. https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/

File: bones/textBone.py [hunk 2]
```diff
@@ -7,108 +7,193 @@
...
-	def handle_entityref(self, name): #FIXME
+	def handle_entityref(self, name):  # FIXME
 		if name in htmlentitydefs.entitydefs.keys():
-			self.result += "&%s;" % ( name )
+			self.flushCache()
+			self.result += "&%s;" % (name)
+
+	def flushCache(self):
+		"""
+			Flush pending tags into the result and push their corresponding end-tags onto the stak
```
Comment (skoegl) [inline bones/textBone.py]: typo 'stack'

File: bones/textBone.py (hunk 1 above)
Comment (tsteinruecken) [reply bones/textBone.py]: That's enforced automatically if target is set to blank (Line 152f)

**Comment by sveneberth** [comment]
What about this pull request? Its necessary for the new HTML Editor of the Vi. See [#17@Vi](https://github.com/viur-framework/vi/pull/17)

**Comment by tsteinruecken** [comment]
I don't know, Up to my knowledge we hasn't reached a stable implementation yet. If it's now working correctly in VI, gracefully handles old formatting and has been implemented in the admin also we can merge it.

**Comment by phorward** [comment]
> If it's now working correctly in VI, gracefully handles old formatting and has been implemented in the admin also we can merge it.

Hello, this pull request is scheduled for ViUR v2.4 release. It will replace the old implementation of the textBone in the VI with an entirely working, stable version. The admin is not tendered since several iterations in the past, so in my opinion this shall be omitted, or someone will update it. The ViUR project cannot stop to continue when old administration tools aren't updated recently. Time is going on.

Anyway, a possibility I'd like to discuss is renaming the textBone to htmlBone, especially because it is a better description of the kind of data that is stored into these kinds of bones. In such a case, the admin can still be used with the old editor, but VI will not continue to hold old things anymore and replaces textBone by the newly developed htmlBone. 

Is this an option?

**Comment by phorward** [comment]
This pull request blocks [vi #17](https://github.com/viur-framework/vi/pull/17) and needs to be merged asap into develop...

**Comment by tsteinruecken** [comment]
Superseded by #149.

**Pull request closed without merging**